toolkit/tools/imagegen/configuration/users.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Parser for the image builder's Users configuration schema.
+
+package configuration
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type User struct {
+	Name                string   `json:"Name"`
+	UID                 string   `json:"UID"`
+	PasswordHashed      bool     `json:"PasswordHashed"`
+	Password            string   `json:"Password"`
+	PasswordExpiresDays int64    `json:"PasswordExpiresDays"`
+	SSHPubKeyPaths      []string `json:"SSHPubKeyPaths"`
+	PrimaryGroup        string   `json:"PrimaryGroup"`
+	SecondaryGroups     []string `json:"SecondaryGroups"`
+	StartupCommand      string   `json:"StartupCommand"`
+}
+
+// IsValid returns an error if the User struct is not valid
+func (p *User) IsValid() (err error) {
+	err = p.NameIsValid()
+	if err != nil {
+		return
+	}
+	err = p.UIDIsValid()
+	if err != nil {
+		return
+	}
+	/* err = p.PasswordIsValid()
+	if err != nil {
+		return
+	} */
+	err = p.PasswordExpiresDaysIsValid()
+	if err != nil {
+		return
+	}
+	return
+}
+
+// NameIsValid returns an error if the User name is empty
+func (p *User) NameIsValid() (err error) {
+	if p.Name == "" {
+		return fmt.Errorf("invalid value for name (%s)", p.Name)
+	}
+	return
+}
+
+// UIDIsValid returns an error if the UID is outside bounds
+// UIDs 1-999 are system users and 1000-60000 are normal users
+// Bounds can be checked using:
+// $grep -E '^UID_MIN|^UID_MAX' /etc/login.defs
+func (p *User) UIDIsValid() (err error) {
+	const (
+		uidLowerBound = 0 // root user
+		uidUpperBound = 60000
+	)
+	if p.UID != "" {
+		uidNum, err := strconv.Atoi(p.UID)
+		if err != nil || uidNum < uidLowerBound || uidNum > uidUpperBound {
+			return fmt.Errorf("invalid value for UID (%s)", p.UID)
```

Comment:
I'd also include the value of `err` in the message or (better yet) print a separate message for it. Example:
- for `err != nil` log an error along the lines of `"Failed to convert user ID (%s) to a number with: %v", err` and then return the value of `err` from the function.
- for UIDs out of bounds return an error saying `"UID (%s) out of bound <%d, %d>", p.UID, uidLowerBound, uidUpperBound`.

Reply:
The pattern I've used for all the other configuration checks uses error wrapping (`return fmt.Errorf("I'm a more general error, I ran into this error while doing my thing: '%w'", err)`. This gives a single error to print at the end (i.e. `SystemConfig.IsValid()` should call `Users.IsValid()`, and wrap the error it gets back and return that.

Generally I feel like "Am I Valid?" type functions shouldn't print errors on bad input since they are functioning exactly as intended. It should be up to the caller to decide what to do with the error.

Reply:
We're mixing validation errors (when the value provided by the user is invalid) and execution errors (when something went wrong in the code but it's not user's fault). For a clearer example, a function checking if a content of a file is valid shouldn't silently (without indicating what is the source of the error) say that it's not just because someone removed the hard drive mid way - it's a separate issue, not the one the function is meant to handle. I think that wrapping the error is fine but we should still keep `err != nil` case separate from the bound checks.

---
In reply to: [532845634](https://github.com/microsoft/CBL-Mariner/pull/419#discussion_r532845634) [](ancestors = 532845634)

Reply:
Ah, fair. Agree with that.

Reply:
Agreed.  Fixed, thanks for the feedback!